File: Sources/Usage/DeviceContext.swift
// The per-event `context` the ingest accepts: coarse facts about the app and the
// device it runs on, gathered once per process.
//
//   Apple      appVersion, osName, osVersion, deviceModel, formFactor, locale
//   Linux      osName, osVersion (kernel major)
//   Android    osName
//   WASI page  browserName, browserVersion, osName, formFactor, locale
//   WASI Node  osName
//
// A server (platform "server", which includes macOS) and any client whose device
// id the caller or host supplied send only osName, a major-only osVersion and
// appVersion: that device is a tenant's, not the host's, so the host's model and
// locale say nothing about it. A native macOS app is in that set only through its
// platform tag; a Catalyst or iOS-on-Mac app reports "ios" and sends the full set.
//
// The ingest rejects a whole batch whose context is 4096 bytes or more, so every
// context is cut down here (`sanitizeContext`) before it can reach a send.

import JSON

#if canImport(Darwin)
import Darwin
import Foundation
#elseif os(Android)
#elseif canImport(Glibc)
import Glibc
#elseif canImport(Musl)
import Musl
#elseif os(WASI)
import JavaScriptKit
#endif

/// The keys the ingest accepts in an event's `context`. Anything else is dropped.
let contextKeys: Set<String> = [
    "appVersion", "osName", "osVersion", "deviceModel",
    "browserName", "browserVersion", "formFactor", "locale",
]

/// The values the ingest accepts for `formFactor`.
let formFactors: Set<String> = ["desktop", "mobile", "tablet"]

/// Per-value cap. The ingest allows 256 characters; nothing we send needs a
/// quarter of that, so a longer value is a host override gone wrong.
let maxContextValueBytes = 64

/// Cap on the encoded context. Well under the ingest's 4096, so one oversized
/// context can never cost the batch; over it, the event goes without context.
let maxContextBytes = 1024

/// Facts about this process's host, gathered once (`current`).
struct DeviceContext: Sendable, Equatable {
    var appVersion: String?
    var osName: String?
    /// major.minor where the platform has one.
    var osVersion: String?
    var deviceModel: String?
    var browserName: String?
    /// Major only.
    var browserVersion: String?
    var formFactor: String?
    /// language-region, e.g. "pt-BR".
    var locale: String?

    static let current = DeviceContext.detect()

    /// The context sent for this host. `minimal` is the server set.
    func fields(minimal: Bool, appVersionOverride: String? = nil) -> [String: String] {
        var out: [String: String] = [:]
        out["appVersion"] = appVersionOverride ?? appVersion
        out["osName"] = osName
        if minimal {
            out["osVersion"] = osVersion.map(majorVersion)
            return out
        }
        out["osVersion"] = osVersion
        out["deviceModel"] = deviceModel
        out["browserName"] = browserName
        out["browserVersion"] = browserVersion
        out["formFactor"] = formFactor
        out["locale"] = locale
        return out
    }
}

/// The default `context` provider `makeClient` wires. The host facts are cached;
/// the opt-outs and the appVersion override are read per event, so a host that
/// sets them after the first client is built is still honoured.
func defaultContextProvider(platform: String, deviceIdSupplied: Bool) -> () -> [String: String]? {
    let minimal = platform == "server" || deviceIdSupplied
    return {
        if deviceContextDisabled() { return nil }
        return DeviceContext.current.fields(minimal: minimal, appVersionOverride: hostProvidedAppVersion())
    }
}

/// `context` reduced to what the ingest will accept without rejecting the batch:
/// allowlisted keys, printable values of at most `maxContextValueBytes`, a known
/// formFactor, and at most `maxContextBytes` encoded. `nil` when nothing is left
/// or the whole is still too big.
func sanitizeContext(_ context: [String: String]?) -> [String: String]? {
    guard let context else { return nil }
    var out: [String: String] = [:]
    for (key, raw) in context where contextKeys.contains(key) {
        let value = printableValue(raw)
        if value.isEmpty { continue }
        if key == "formFactor" && !formFactors.contains(value) { continue }
        out[key] = value
    }
    if out.isEmpty { return nil }
    guard let json = try? JSONEncoder().encodeToString(out), json.utf8.count <= maxContextBytes else {
        return nil
    }
    return out
}

/// `raw` without control or invisible formatting characters, trimmed, and cut
/// to `maxContextValueBytes` on a character boundary.
func printableValue(_ raw: String) -> String {
    var kept = String.UnicodeScalarView()
    for scalar in raw.unicodeScalars where isPrintable(scalar) { kept.append(scalar) }
    var out = ""
    var bytes = 0
    for character in trimmed(String(kept)) {
        bytes += character.utf8.count
        if bytes > maxContextValueBytes { break }
        out.append(character)
    }
    return trimmed(out)
}

private func isPrintable(_ scalar: Unicode.Scalar) -> Bool {
    switch scalar.value {
    case 0..<0x20, 0x7F...0x9F: return false            // C0, DEL, C1
    case 0x200B...0x200F, 0x2028...0x202E, 0x2060...0x206F: return false // zero-width, separators, bidi
    case 0xAD, 0x180E, 0xFE00...0xFE0F, 0xFEFF, 0xFFF9...0xFFFB, 0xE0000...0xE007F: return false
    default: return true
    }
}

private func trimmed(_ value: String) -> String {
    String(value.drop(while: \.isWhitespace).reversed().drop(while: \.isWhitespace).reversed())
}

/// "15.6" -> "15". The server set's coarse version.
func majorVersion(_ version: String) -> String {
    String(version.prefix(while: { $0 != "." }))
}

/// A BCP 47 tag reduced to language and region: "zh-Hant-TW" -> "zh-TW",
/// "en_US" -> "en-US", "fr" -> "fr". `nil` when it does not start with a language.
func languageRegion(_ tag: String?) -> String? {
    guard let tag else { return nil }
    let parts = tag.split(whereSeparator: { $0 == "-" || $0 == "_" || $0 == "@" || $0 == "." })
    guard let first = parts.first, (2...3).contains(first.count), first.allSatisfy(\.isASCIILetter) else {
        return nil
    }
    let language = first.lowercased()
    for part in parts.dropFirst() {
        if part.count == 2, part.allSatisfy(\.isASCIILetter) { return language + "-" + part.uppercased() }
        if part.count == 3, part.allSatisfy(\.isASCIIDigit) { return language + "-" + part }
        // The region follows the script; an extension or a variant means none came.
        if part.count != 4 { break }
    }
    return language
}

private extension Character {
    var isASCIILetter: Bool { ("a"..."z").contains(self) || ("A"..."Z").contains(self) }
    var isASCIIDigit: Bool { ("0"..."9").contains(self) }
}

// MARK: - Browser

/// The browser vocabulary the dashboard groups by.
let browserNames: Set<String> = ["Chrome", "Edge", "Safari", "Firefox", "Opera", "Samsung Internet", "Other"]

/// Browser name and major version. User-Agent Client Hints first (Chromium
/// browsers), then the user agent string, which is all Safari and Firefox have.
/// A Chromium browser whose brands name none we know (Brave, Vivaldi) is
/// "Other": its user agent string claims Chrome.
func browserIdentity(brands: [(brand: String, version: String)], userAgent: String) -> (name: String, version: String?) {
    let named = brands.filter { !$0.brand.contains("Brand") && $0.brand != "Chromium" }
    if !brands.isEmpty {
        for (prefix, name) in [("Microsoft Edge", "Edge"), ("Opera", "Opera"),
                               ("Samsung Internet", "Samsung Internet"), ("Google Chrome", "Chrome")] {
            if let hit = named.first(where: { $0.brand.hasPrefix(prefix) }) {
                return (name, leadingDigits(hit.version))
            }
        }
        return ("Other", nil)
    }
    for (token, name) in [
        ("SamsungBrowser/", "Samsung Internet"),
        ("OPR/", "Opera"), ("OPiOS/", "Opera"), ("OPT/", "Opera"),
        ("Edg/", "Edge"), ("EdgA/", "Edge"), ("EdgiOS/", "Edge"), ("Edge/", "Edge"),
        ("FxiOS/", "Firefox"), ("Firefox/", "Firefox"),
        ("CriOS/", "Chrome"), ("Chrome/", "Chrome"),
    ] {
        if let version = versionAfter(token, in: userAgent) { return (name, version) }
    }
    if userAgent.contains("Safari/"), let version = versionAfter("Version/", in: userAgent) {
        return ("Safari", version)
    }
    return ("Other", nil)
}

/// The OS a page runs on. Client Hints' platform first, then the user agent.
/// iPadOS 13+ Safari sends a Mac user agent; touch points give it away.
func browserOSName(hintPlatform: String?, userAgent: String, maxTouchPoints: Int) -> String? {
    switch hintPlatform {
    case "Windows": return "Windows"
    case "macOS": return maxTouchPoints > 1 ? "iPadOS" : "macOS"
    case "Linux": return "Linux"
    case "Android": return "Android"
    case "Chrome OS", "ChromeOS": return "ChromeOS"
    case "iOS": return "iOS"
    default: break
    }
    if userAgent.contains("iPad") { return "iPadOS" }
    if userAgent.contains("iPhone") || userAgent.contains("iPod") { return "iOS" }
    if userAgent.contains("Android") { return "Android" }
    if userAgent.contains("CrOS") { return "ChromeOS" }
    if userAgent.contains("Windows") { return "Windows" }
    if userAgent.contains("Macintosh") || userAgent.contains("Mac OS X") {
        return maxTouchPoints > 1 ? "iPadOS" : "macOS"
    }
    if userAgent.contains("Linux") { return "Linux" }
    return nil
}

/// Always one of `formFactors`. An Android user agent without "Mobile" is a
/// tablet by Google's own convention, and a Mac user agent with touch is an iPad.
func browserFormFactor(userAgent: String, mobileHint: Bool?, maxTouchPoints: Int) -> String {
    if userAgent.contains("iPad") { return "tablet" }
    if userAgent.contains("Macintosh") && maxTouchPoints > 1 { return "tablet" }
    if userAgent.contains("iPhone") || userAgent.contains("iPod") { return "mobile" }
    if userAgent.contains("Android") { return userAgent.contains("Mobile") ? "mobile" : "tablet" }
    if mobileHint == true { return "mobile" }
    return "desktop"
}

/// Node's `process.platform` in the vocabulary the other hosts use.
func nodeOSName(_ platform: String?) -> String? {
    switch platform {
    case nil, "": return nil
    case "darwin": return "macOS"
    case "linux": return "Linux"
    case "win32": return "Windows"
    case "android": return "Android"
    case let other?: return other
    }
}

private func versionAfter(_ token: String, in userAgent: String) -> String? {
    guard let range = userAgent.firstRange(of: token) else { return nil }
    return leadingDigits(String(userAgent[range.upperBound...]))
}

private func leadingDigits(_ value: String) -> String? {
    let digits = value.prefix(while: \.isASCIIDigit)
    return digits.isEmpty ? nil : String(digits)
}

// MARK: - Detection

extension DeviceContext {
    static func detect() -> DeviceContext {
#if canImport(Darwin)
        return appleContext()
#elseif os(Linux)
        return DeviceContext(osName: "Linux", osVersion: unameRelease().map(majorMinor))
#elseif os(Android)
        return DeviceContext(osName: "Android")
#elseif os(WASI)
        return jsHostIsNode() ? nodeContext() : browserContext()
#else
        return DeviceContext()
#endif
    }
}

/// "6.8.0-45-generic" -> "6.8".
private func majorMinor(_ version: String) -> String {
    let numeric = version.prefix(while: { $0.isASCIIDigit || $0 == "." })
    return numeric.split(separator: ".").prefix(2).joined(separator: ".")
}

#if os(Linux)
private func unameRelease() -> String? {
    var name = utsname()
    guard uname(&name) == 0 else { return nil }
    return withUnsafeBytes(of: &name.release) { String(decoding: $0.prefix(while: { $0 != 0 }), as: UTF8.self) }
}
#endif

#if canImport(Darwin)
private func appleContext() -> DeviceContext {
    let info = ProcessInfo.processInfo
    var context = DeviceContext(
        appVersion: Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String,
        locale: currentLocale()
    )
    let version = info.operatingSystemVersion
    let processVersion = "\(version.majorVersion).\(version.minorVersion)"
#if os(macOS)
    context.osName = "macOS"
    context.osVersion = processVersion
    context.deviceModel = sysctlString("hw.model")
    context.formFactor = "desktop"
#elseif os(iOS)
    // Catalyst and iOS apps on a Mac count as macOS: the hardware is a Mac, and
    // their ProcessInfo reports an iOS version number that no Mac ever ran.
    let onMac = info.isMacCatalystApp || info.isiOSAppOnMac
    if onMac {
        context.osName = "macOS"
        context.osVersion = sysctlString("kern.osproductversion").map(majorMinor)
        context.deviceModel = sysctlString("hw.model")
        context.formFactor = "desktop"
    } else {
#if targetEnvironment(simulator)
        let model = info.environment["SIMULATOR_MODEL_IDENTIFIER"]
#else
        let model = machineIdentifier()
#endif
        let iPad = model?.hasPrefix("iPad") == true
        let phone = model.map { $0.hasPrefix("iPhone") || $0.hasPrefix("iPod") } == true
        context.osName = iPad ? "iPadOS" : "iOS"
        context.osVersion = processVersion
        context.deviceModel = model
        context.formFactor = iPad ? "tablet" : phone ? "mobile" : nil
    }
#else
    // tvOS, visionOS, watchOS: none of the three form factors fits.
#if os(tvOS)
    context.osName = "tvOS"
#elseif os(visionOS)
    context.osName = "visionOS"
#elseif os(watchOS)
    context.osName = "watchOS"
#endif
    context.osVersion = processVersion
#if targetEnvironment(simulator)
    context.deviceModel = info.environment["SIMULATOR_MODEL_IDENTIFIER"]
#else
    context.deviceModel = machineIdentifier()
#endif
#endif
    return context
}

private func currentLocale() -> String? {
    guard let language = Locale.current.language.languageCode?.identifier else { return nil }
    return languageRegion(Locale.current.region.map { "\(language)-\($0.identifier)" } ?? language)
}

/// `utsname.machine`, e.g. "iPhone16,2". Read directly rather than through
/// UIDevice, which is MainActor-bound and has no model identifier anyway.
private func machineIdentifier() -> String? {
    var name = utsname()
    guard uname(&name) == 0 else { return nil }
    let value = withUnsafeBytes(of: &name.machine) { String(decoding: $0.prefix(while: { $0 != 0 }), as: UTF8.self) }
    return value.isEmpty ? nil : value
}

private func sysctlString(_ name: String) -> String? {
    var size = 0
    guard sysctlbyname(name, nil, &size, nil, 0) == 0, size > 0 else { return nil }
    var buffer = [UInt8](repeating: 0, count: size)
    guard sysctlbyname(name, &buffer, &size, nil, 0) == 0 else { return nil }
    let value = String(decoding: buffer.prefix(while: { $0 != 0 }), as: UTF8.self)
    return value.isEmpty ? nil : value
}
#endif

#if os(WASI)
private func nodeContext() -> DeviceContext {
    DeviceContext(osName: nodeOSName(JSObject.global.process.object?.platform.string))
}

private func browserContext() -> DeviceContext {
    guard let navigator = JSObject.global.navigator.object else { return DeviceContext() }
    let userAgent = navigator.userAgent.string ?? ""
    let touchPoints = Int(navigator.maxTouchPoints.number ?? 0)
    let hints = navigator.userAgentData.object
    var brands: [(brand: String, version: String)] = []
    if let list = hints?.brands.object, let count = list.length.number {
        for index in 0..<Int(count) {
            guard let entry = list[index].object else { continue }
            brands.append((entry.brand.string ?? "", entry.version.string ?? ""))
        }
    }
    let browser = browserIdentity(brands: brands, userAgent: userAgent)
    return DeviceContext(
        osName: browserOSName(hintPlatform: hints?.platform.string, userAgent: userAgent, maxTouchPoints: touchPoints),
        browserName: browser.name,
        browserVersion: browser.version,
        formFactor: browserFormFactor(userAgent: userAgent, mobileHint: hints?.mobile.boolean, maxTouchPoints: touchPoints),
        locale: languageRegion(navigator.language.string)
    )
}
#endif

File: packages/tongue-kotlin/src/main/kotlin/ai/desertant/tongue/usage/DeviceContext.kt
package ai.desertant.tongue.usage

import java.text.BreakIterator
import java.util.Locale

/**
 * The per-event `context`: coarse facts about the app and the device it runs on.
 * A port of desert-ant-core's `Sources/Usage/DeviceContext.swift`, which sets the
 * rules; core reads Android's facts through its Kotlin host bridge, and this port
 * reads the same ones reflectively (see Storage.kt for why).
 *
 *   Android   appVersion, osName, osVersion, deviceModel, formFactor, locale
 *   JVM       osName, a major-only osVersion (a JVM is a `server`)
 *
 * A server, and any client whose device id the host supplied, sends only osName,
 * a major-only osVersion and appVersion: that device is not this process's to
 * describe. Nothing that identifies the device: no serial, ANDROID_ID or build
 * fingerprint. `DAL_APP_VERSION` overrides the app version, and
 * `DesertAnt.sendsDeviceContext = false` or `DAL_USAGE_CONTEXT_DISABLED` sends
 * usage without any context.
 */

/** The keys the ingest accepts in an event's `context`. Anything else is dropped. */
internal val CONTEXT_KEYS: Set<String> = setOf(
    "appVersion", "osName", "osVersion", "deviceModel",
    "browserName", "browserVersion", "formFactor", "locale",
)

/** The values the ingest accepts for `formFactor`. */
internal val FORM_FACTORS: Set<String> = setOf("desktop", "mobile", "tablet")

/** Per-value cap, in UTF-8 bytes. Core's, so every port reports the same values. */
internal const val MAX_CONTEXT_VALUE_BYTES: Int = 64

/**
 * Cap on the encoded context. The ingest rejects the whole batch at 4096 bytes,
 * so this stays well under it; over it, the event goes without context.
 */
internal const val MAX_CONTEXT_BYTES: Int = 1024

/** Facts about this host, read once per client. */
internal data class DeviceFacts(
    val appVersion: String? = null,
    val osName: String? = null,
    /** major.minor where the platform has one. */
    val osVersion: String? = null,
    val deviceModel: String? = null,
    val formFactor: String? = null,
    /** language-region, e.g. "pt-BR". */
    val locale: String? = null,
) {
    /** The context sent for this host. `minimal` is the server set. */
    fun fields(minimal: Boolean, appVersionOverride: String? = null): Map<String, String> {
        val out = linkedMapOf<String, String?>()
        out["appVersion"] = appVersionOverride ?: appVersion
        out["osName"] = osName
        if (minimal) {
            out["osVersion"] = osVersion?.substringBefore('.')
        } else {
            out["osVersion"] = osVersion
            out["deviceModel"] = deviceModel
            out["formFactor"] = formFactor
            out["locale"] = locale
        }
        return out.filterValues { it != null }.mapValues { it.value!! }
    }
}

/**
 * The default `context` provider [makeClient] wires. The facts are read once;
 * the opt-out and the appVersion override are read per event, so a host that
 * sets either after the client is built is still honoured.
 */
internal fun defaultContextProvider(
    platform: String,
    deviceIdSupplied: Boolean,
    facts: DeviceFacts,
): () -> Map<String, String>? {
    val minimal = platform == "server" || deviceIdSupplied
    return { facts.fields(minimal, hostProvidedAppVersion()) }
}

/** `DAL_APP_VERSION`, the environment then the system property, or null. */
internal fun hostProvidedAppVersion(): String? = setting("DAL_APP_VERSION")

/**
 * Whether the event `context` is switched off: `DesertAnt.sendsDeviceContext`
 * set to false in code, or `DAL_USAGE_CONTEXT_DISABLED` under `flagIsSet` in the
 * environment or as the same-named system property. Either one opts out.
 * Usage itself still reports; only the context goes.
 */
internal fun deviceContextDisabled(): Boolean =
    !ai.desertant.tongue.DesertAnt.sendsDeviceContext ||
        flagIsSet(readEnvironment("DAL_USAGE_CONTEXT_DISABLED")) ||
        flagIsSet(System.getProperty("DAL_USAGE_CONTEXT_DISABLED"))

/**
 * The truthiness rule every port's opt-out flags share, usage and context alike:
 * set, and not "", "0" or "false". This port reads strings only (the environment
 * and system properties), so a number arrives as its text: "1" is set and "0"
 * is not, as the JS ports' number rule has it. It fails closed.
 */
internal fun flagIsSet(value: String?): Boolean =
    value != null && value != "" && value != "0" && value != "false"

/**
 * `context` reduced to what the ingest will accept without rejecting the batch:
 * allowlisted keys, printable values of at most [MAX_CONTEXT_VALUE_BYTES], a
 * known formFactor, and at most [MAX_CONTEXT_BYTES] encoded. Null when nothing
 * is left or the whole is still too big.
 */
internal fun sanitizeContext(context: Map<String, String>?): Map<String, String>? {
    if (context == null) return null
    val out = linkedMapOf<String, String>()
    for ((key, raw) in context) {
        if (key !in CONTEXT_KEYS) continue
        val value = printableValue(raw)
        if (value.isEmpty()) continue
        if (key == "formFactor" && value !in FORM_FACTORS) continue
        out[key] = value
    }
    if (out.isEmpty()) return null
    val encoded = buildString { appendContext(out) }
    return if (encoded.toByteArray(Charsets.UTF_8).size <= MAX_CONTEXT_BYTES) out else null
}

/**
 * `raw` without control or invisible formatting characters, trimmed, and cut to
 * [MAX_CONTEXT_VALUE_BYTES] without splitting a character (a grapheme, as core
 * cuts on Swift's `Character`).
 */
internal fun printableValue(raw: String): String {
    val kept = buildString {
        var index = 0
        while (index < raw.length) {
            val code = raw.codePointAt(index)
            if (isPrintable(code)) appendCodePoint(code)
            index += Character.charCount(code)
        }
    }.trim()
    val out = StringBuilder()
    var bytes = 0
    val breaks = BreakIterator.getCharacterInstance(Locale.ROOT)
    breaks.setText(kept)
    var start = breaks.first()
    var end = breaks.next()
    while (end != BreakIterator.DONE) {
        val character = kept.substring(start, end)
        bytes += character.toByteArray(Charsets.UTF_8).size
        if (bytes > MAX_CONTEXT_VALUE_BYTES) break
        out.append(character)
        start = end
        end = breaks.next()
    }
    return out.toString().trim()
}

private fun isPrintable(code: Int): Boolean = when (code) {
    in 0 until 0x20, in 0x7F..0x9F -> false // C0, DEL, C1
    in 0xD800..0xDFFF -> false // a lone surrogate, which a strict JSON reader rejects
    in 0x200B..0x200F, in 0x2028..0x202E, in 0x2060..0x206F -> false // zero-width, separators, bidi
    0xAD, 0x180E, in 0xFE00..0xFE0F, 0xFEFF, in 0xFFF9..0xFFFB, in 0xE0000..0xE007F -> false
    else -> true
}

/** "8.1.0" -> "8.1", "6.8.0-45-generic" -> "6.8". Null when it does not start with a number. */
internal fun majorMinor(version: String): String? {
    val numeric = version.trim().takeWhile { it.isDigit() || it == '.' }
    return numeric.split('.').filter { it.isNotEmpty() }.take(2).joinToString(".").ifEmpty { null }
}

/** Google's own tablet line: a smallest width of 600dp or more. 0 is undefined. */
internal fun formFactor(smallestWidthDp: Int): String? = when {
    smallestWidthDp <= 0 -> null
    smallestWidthDp >= 600 -> "tablet"
    else -> "mobile"
}

/**
 * Language and region only: "zh-Hant-TW" -> "zh-TW", "fr" -> "fr". Through the
 * language tag, so Hebrew reads "he" rather than the legacy "iw".
 */
internal fun languageRegion(locale: Locale): String? {
    val language = locale.toLanguageTag().substringBefore('-')
    if (language.isEmpty() || language == "und") return null
    return if (locale.country.isEmpty()) language else "$language-${locale.country}"
}

/**
 * The facts from each reader. A reader that throws, or returns nothing usable,
 * leaves its fact out rather than failing the model's construction.
 */
internal fun androidFacts(
    appVersion: () -> String?,
    osRelease: () -> String?,
    model: () -> String?,
    smallestWidthDp: () -> Int?,
    locale: () -> Locale?,
): DeviceFacts = DeviceFacts(
    appVersion = attempt(appVersion)?.trim()?.ifEmpty { null },
    osName = "Android",
    osVersion = attempt(osRelease)?.let(::majorMinor),
    deviceModel = attempt(model)?.trim()?.ifEmpty { null },
    formFactor = attempt(smallestWidthDp)?.let(::formFactor),
    locale = attempt(locale)?.let(::languageRegion),
)

private fun <T> attempt(read: () -> T?): T? = try { read() } catch (_: Throwable) { null }

/**
 * This host's facts: Android's through reflection when it is Android (with or
 * without a Context), otherwise the JVM's OS name and version.
 */
internal fun detectDeviceFacts(context: Any?): DeviceFacts =
    if (isAndroid()) reflectedAndroidFacts(context) else jvmFacts()

/**
 * Android's facts, reached reflectively so this file compiles and runs on a bare
 * JVM. Methods are looked up on the public SDK classes rather than the runtime
 * class, which may be a hidden one reflection cannot call. Without a Context
 * there is no app version, and the form factor comes from the system resources.
 */
private fun reflectedAndroidFacts(context: Any?): DeviceFacts {
    val contextClass = runCatching { Class.forName("android.content.Context") }.getOrNull()
    val appContext = context?.takeIf { contextClass?.isInstance(it) == true }
    // The application's configuration, not an Activity's: a tablet Activity in
    // split screen can report under 600dp, and the facts last the client.
    val application = appContext?.let { runCatching { contextClass!!.getMethod("getApplicationContext").invoke(it) }.getOrNull() }
        ?: appContext
    return androidFacts(
        appVersion = {
            appContext?.let {
                val packageManager = contextClass!!.getMethod("getPackageManager").invoke(it)
                val packageName = contextClass.getMethod("getPackageName").invoke(it)
                val info = Class.forName("android.content.pm.PackageManager")
                    .getMethod("getPackageInfo", String::class.java, Int::class.javaPrimitiveType)
                    .invoke(packageManager, packageName, 0)
                Class.forName("android.content.pm.PackageInfo").getField("versionName").get(info) as String?
            }
        },
        osRelease = { Class.forName("android.os.Build\$VERSION").getField("RELEASE").get(null) as String? },
        model = { Class.forName("android.os.Build").getField("MODEL").get(null) as String? },
        smallestWidthDp = {
            val resourcesClass = Class.forName("android.content.res.Resources")
            val resources = application?.let { contextClass!!.getMethod("getResources").invoke(it) }
                ?: resourcesClass.getMethod("getSystem").invoke(null)
            val configuration = resourcesClass.getMethod("getConfiguration").invoke(resources)
            Class.forName("android.content.res.Configuration").getField("smallestScreenWidthDp").getInt(configuration)
        },
        locale = {
            val localeList = Class.forName("android.os.LocaleList")
            val defaults = localeList.getMethod("getDefault").invoke(null)
            localeList.getMethod("get", Int::class.javaPrimitiveType).invoke(defaults, 0) as Locale?
        },
    )
}

/** A JVM is a `server`, which sends the OS and a major-only version. */
private fun jvmFacts(): DeviceFacts {
    val name = System.getProperty("os.name").orEmpty()
    val osName = when {
        name.startsWith("Mac") -> "macOS"
        name.startsWith("Windows") -> "Windows"
        name.startsWith("Linux") -> "Linux"
        else -> name.ifEmpty { null }
    }
    return DeviceFacts(osName = osName, osVersion = System.getProperty("os.version")?.let(::majorMinor))
}
